exporter/splunkhecexporter/client.go

```diff
@@ -36,25 +37,156 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk"
 )
 
+var (
+	errOverCapacity = errors.New("over capacity")
+)
+
+// Minimum number of bytes to compress. 1500 is the MTU of an ethernet frame.
+const minCompressionLen = 1500
+
 // client sends the data to the splunk backend.
 type client struct {
-	config  *Config
-	url     *url.URL
-	client  *http.Client
-	logger  *zap.Logger
-	zippers sync.Pool
-	wg      sync.WaitGroup
-	headers map[string]string
+	config         *Config
+	url            *url.URL
+	client         *http.Client
+	logger         *zap.Logger
+	wg             sync.WaitGroup
+	headers        map[string]string
+	gzipWriterPool *sync.Pool
 }
 
 // bufferState encapsulates intermediate buffer state when pushing data
 type bufferState struct {
-	buf      *bytes.Buffer
-	tmpBuf   *bytes.Buffer
-	bufFront *index
-	bufLen   int
-	resource int
-	library  int
+	compressionAvailable bool
+	compressionEnabled   bool
+	bufferMaxLen         uint
+	writer               io.Writer
+	buf                  *bytes.Buffer
+	bufFront             *index
+	resource             int
+	library              int
+	gzipWriterPool       *sync.Pool
+}
+
+func (b *bufferState) reset() {
+	b.buf.Reset()
+	b.compressionEnabled = false
+	b.writer = &cancellableBytesWriter{innerWriter: b.buf, maxCapacity: b.bufferMaxLen}
+}
+
+func (b *bufferState) Read(p []byte) (n int, err error) {
+	return b.buf.Read(p)
+}
+
+func (b *bufferState) Close() error {
+	if _, ok := b.writer.(*cancellableGzipWriter); ok {
+		return b.writer.(*cancellableGzipWriter).close()
+	}
+	return nil
+}
+
+// accept returns true if data is accepted by the buffer
+func (b *bufferState) accept(data []byte) bool {
+	_, err := b.writer.Write(data)
+	overCapacity := errors.Is(err, errOverCapacity)
+	bufLen := b.buf.Len()
+	if overCapacity {
+		bufLen += len(data)
+	}
+	if b.compressionAvailable && !b.compressionEnabled && bufLen > minCompressionLen {
+		// switch over to a zip buffer.
+		tmpBuf := bytes.NewBuffer(make([]byte, 0, b.bufferMaxLen+bufCapPadding))
+		writer := b.gzipWriterPool.Get().(*gzip.Writer)
+		writer.Reset(tmpBuf)
+		zipWriter := &cancellableGzipWriter{
+			innerBuffer: tmpBuf,
+			innerWriter: writer,
+			// 8 bytes required for the zip footer.
+			maxCapacity:    b.bufferMaxLen - 8,
+			gzipWriterPool: b.gzipWriterPool,
+		}
+
+		// the new data is so big, even with a zip writer, we are over the max limit.
+		// abandon and return false, so we can send what is already in our buffer.
+		if _, err2 := zipWriter.Write(b.buf.Bytes()); err2 != nil {
+			return false
+		}
+		b.writer = zipWriter
+		b.buf = tmpBuf
+		b.compressionEnabled = true
+		// if the byte writer was over capacity, try to write the new entry in the zip writer:
+		if overCapacity {
+			if _, err2 := zipWriter.Write(data); err2 != nil {
+				return false
```

Comment:
I believe we logged `dropped log event` in this case before. Should we keep that logging?

Reply:
We still log line 409, 464 and 514. We use this terminology:
```
permanentErrors = append(permanentErrors, consumererror.NewPermanent(
				fmt.Errorf("dropped log event error: event size %d bytes larger than configured max content length %d bytes", len(b), state.bufferMaxLen)))
```